### PR TITLE
Refactor `previewctl install-context`

### PR DESCRIPTION
**Pull request by vulkoingim**
## Description

* Changes the underlying implementation of the `previewctl install-context` command from bash to go and adds tests
* Changes parts of the `get-credentials` command, that was introduced recently so it fits better with the new design and adds tests to some of the functions as well
* It doesn't change how preivewctl is used and the context is installed in workspaces yet (follow up pr will deal with that) 

## Related Issue(s)

Fixes https://github.com/gitpod-io/ops/issues/5763

## How to test

```bash
gcloud auth login

cd dev/preview/previewctl 

KUBECONFIG=$(pwd)/tmp go run main.go install-context
```

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by roboquat** [comment]
@vulkoingim: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: dev/preview/previewctl/pkg/k8s/context.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+import (
+	"fmt"
+
+	"github.com/imdario/mergo"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+)
+
+func RenameConfig(config *api.Config, oldName, newName string) (*api.Config, error) {
+	// TODO: add tests
```
Comment (mads-hartmann) [inline dev/preview/previewctl/pkg/k8s/context.go:16]: Nit: Lets just remove TODOs like this

File: dev/preview/previewctl/pkg/k8s/context.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	}
+
+	config.AuthInfos[newName] = auth
+	delete(config.AuthInfos, oldName)
+
+	return config, nil
+}
+
+func MergeContextsWithDefault(configs ...*api.Config) (*api.Config, error) {
+	// TODO: add tests
```
Comment (mads-hartmann) [inline dev/preview/previewctl/pkg/k8s/context.go:66]: Nit: Remove comment

File: dev/preview/previewctl/pkg/k8s/context/k3s/k3s.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+	if err := k.client.Close(); err != nil {
+		return err
+	}
+
+	k.client = nil
+	return nil
+}
+
+func (k *ConfigLoader) getVMPodName(ctx context.Context, previewName, namespace string) (string, error) {
+	// TODO: replace this with a call to SVC.Proxy and get the pod name from there
```
Comment (mads-hartmann) [inline dev/preview/previewctl/pkg/k8s/context/k3s/k3s.go:233]: Nit: Let's create a tech-debt issue for this and delete this comment

File: dev/preview/previewctl/pkg/preview/preview.go [hunk 4]
```diff
@@ -138,16 +164,35 @@ func ensureVMICreationTime(p *Preview) {
...
+
+	return k8s.OutputContext(opts.KubeSavePath, merged)
+}
+
+func (p *Preview) GetPreviewContext(ctx context.Context) (*api.Config, error) {
+	return p.configLoader.Load(ctx)
+}
+
+func InstallVMSSHKeys() error {
+	// TODO: reimplement in go
```
Comment (mads-hartmann) [inline dev/preview/previewctl/pkg/preview/preview.go:194]: Nit: Remove TODO and create a tech-debt issue instead

**Review by mads-hartmann — COMMENTED** [review]
This is awesome Aleks. I left a few inline comments.

For a big PR like this I would have loved a PR description that provided more context about the change. Here are a few questions I have that would have been nice to have in the PR description.

1. We'll have to bump the version of previewctl in the image (unfortunately). This isn't done in this PR so I'm assuming you'll do it in a follow-up PR.
1. With this change we should be able to delete [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.gitpod.yml#L41-L44)
1. Install-contexts isn't used in any of the Werft jobs yet. What's the plan for using it? It would be nice to be able to delete [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/dev/preview/util/download-and-merge-harvester-kubeconfig.sh) and switch any invocations of it like [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.werft/ide-integration-tests-startup.sh#L156-L158) to use install-context. We should also be able to simplify the Werft job by removing the stateful handling of port-forwarding that install-context now takes care of (e.g. removing [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.werft/build.ts#L47-L48) and related code)

**Comment by vulkoingim** [comment]
> For a big PR like this I would have loved a PR description that provided more context about the change. Here are a few questions I have that would have been nice to have in the PR description.

Yeah I agree - I just marked it ready for review when I was mostly done with it yesterday and left the description for today. Didn't think that anyone will review it that soon 😄 

> 1. We'll have to bump the version of previewctl in the image (unfortunately). This isn't done in this PR so I'm assuming you'll do it in a follow-up PR.
Yep

> 2. With this change we should be able to delete [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.gitpod.yml#L41-L44)


Also yes


> 3. Install-contexts isn't used in any of the Werft jobs yet. What's the plan for using it? It would be nice to be able to delete [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/dev/preview/util/download-and-merge-harvester-kubeconfig.sh) and switch any invocations of it like [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.werft/ide-integration-tests-startup.sh#L156-L158) to use install-context. We should also be able to simplify the Werft job by removing the stateful handling of port-forwarding that install-context now takes care of (e.g. removing [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.werft/build.ts#L47-L48) and related code)

> It would be nice to be able to delete  [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/dev/preview/util/download-and-merge-harvester-kubeconfig.sh)

Will remove this eventually
> switch any invocations of it like [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.werft/ide-integration-tests-startup.sh#L156-L158)

Will change this later as well

> We should also be able to simplify the Werft job by removing the stateful handling of port-forwarding that install-context now takes care of (e.g. removing [this](https://github.com/gitpod-io/gitpod/blob/172bf96007c979ca18c54d46c307a4aa7e9c2cbc/.werft/build.ts#L47-L48) and related code)

Yeah, we should be able to remove this and the part that copies the context in the job also.

File: dev/preview/previewctl/pkg/k8s/context.go (hunk 1 above)
Comment (vulkoingim) [reply dev/preview/previewctl/pkg/k8s/context.go]: Replaced them with a link to https://github.com/gitpod-io/ops/issues/6524

**Review by mads-hartmann — APPROVED** [review]
🛹 🚀

**Pull request merged**